gno.land/pkg/gnoland/app_test.go

```diff
@@ -683,6 +699,192 @@ func TestEndBlocker(t *testing.T) {
 			assert.Equal(t, changes[index].Power, update.Power)
 		}
 	})
+
+	t.Run("negative power filtered out", func(t *testing.T) {
+		var (
+			key1 = getDummyKey(t)
+			key2 = getDummyKey(t)
+
+			validUpdate = abci.ValidatorUpdate{
+				Address: key1.PubKey().Address(),
+				PubKey:  key1.PubKey(),
+				Power:   1,
+			}
+
+			invalidUpdate = abci.ValidatorUpdate{
+				Address: key2.PubKey().Address(),
+				PubKey:  key2.PubKey(),
+				Power:   -1, // Invalid negative power
+			}
+
+			updates = []abci.ValidatorUpdate{validUpdate, invalidUpdate}
+
+			mockEventSwitch = newCommonEvSwitch()
+
+			mockVMKeeper = &mockVMKeeper{
+				queryFn: func(_ sdk.Context, pkgPath, expr string) (string, error) {
+					require.Equal(t, valRealm, pkgPath)
+					require.NotEmpty(t, expr)
+
+					return constructVMResponse(updates), nil
+				},
+			}
+
+			vmEvents = []abci.Event{
+				gnostd.GnoEvent{
+					Type:    validatorAddedEvent,
+					PkgPath: valRealm,
+				},
+				gnostd.GnoEvent{
+					Type:    validatorAddedEvent,
+					PkgPath: valRealm,
+				},
+			}
+			txEvent = bft.EventTx{
+				Result: bft.TxResult{
+					Response: abci.ResponseDeliverTx{
+						ResponseBase: abci.ResponseBase{
+							Events: vmEvents,
+						},
+					},
+				},
+			}
+		)
+
+		c := newCollector[validatorUpdate](mockEventSwitch, validatorEventFilter)
+		mockEventSwitch.FireEvent(txEvent)
+
+		eb := EndBlocker(c, nil, nil, mockVMKeeper, &mockEndBlockerApp{})
+		res := eb(sdk.Context{}.WithConsensusParams(&abci.ConsensusParams{
+			Validator: &abci.ValidatorParams{
+				PubKeyTypeURLs: []string{"/tm.PubKeySecp256k1"},
+			},
+		}), abci.RequestEndBlock{})
+		require.Len(t, res.ValidatorUpdates, 1)
+		assert.Equal(t, validUpdate.Address, res.ValidatorUpdates[0].Address)
+		assert.Equal(t, validUpdate.Power, res.ValidatorUpdates[0].Power)
+	})
+
+	t.Run("pubkey address mismatch filtered out", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			key1 = getDummyKey(t)
+			key2 = getDummyKey(t)
+			key3 = getDummyKey(t)
```

Comment:
Same comment as before, or use a helper that returns a slice of keys

Reply:
[216b4d9](https://github.com/gnolang/gno/pull/3925/commits/216b4d9cad8b9f7cb6d9cd8cfb0c8cba9e39f38b)